From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Alessandro Gatti <a.gatti@frob.it>
Date: Thu, 16 Jan 2020 00:54:01 +0100
Subject: [PATCH] 1437: Add instructionEndian to generated .ldefs files.

---
 GhidraBuild/Skeleton/data/languages/skel.ldefs | 1 +
 1 file changed, 1 insertion(+)

diff --git a/GhidraBuild/Skeleton/data/languages/skel.ldefs b/GhidraBuild/Skeleton/data/languages/skel.ldefs
index 8cb0d2ac4..f90deb5e2 100644
--- a/GhidraBuild/Skeleton/data/languages/skel.ldefs
+++ b/GhidraBuild/Skeleton/data/languages/skel.ldefs
@@ -7,6 +7,7 @@
 <!-- 
    <language processor="Skel"
             endian="little"
+            instructionEndian="little"
             size="16"
             variant="default"
             version="1.0"
-- 
2.43.0

