diff --git a/Ghidra/Features/Base/src/main/java/ghidra/app/plugin/core/datamgr/archive/DataTypeManagerHandler.java b/Ghidra/Features/Base/src/main/java/ghidra/app/plugin/core/datamgr/archive/DataTypeManagerHandler.java
index 7bb8ea5c2f6..66f8debd23b 100644
--- a/Ghidra/Features/Base/src/main/java/ghidra/app/plugin/core/datamgr/archive/DataTypeManagerHandler.java
+++ b/Ghidra/Features/Base/src/main/java/ghidra/app/plugin/core/datamgr/archive/DataTypeManagerHandler.java
@@ -627,6 +627,11 @@ void dataTypeManagerChanged(FileArchive archive, DataTypeManager oldManager,
 		newManager.addDataTypeManagerListener(listenerDelegate);
 		dataTypeIndexer.removeDataTypeManager(oldManager);
 		dataTypeIndexer.addDataTypeManager(newManager);
+		
+		if (newManager instanceof FileDataTypeManager) {
+			((FileDataTypeManager)newManager).setDataOrganization(plugin.getProgram().getCompilerSpec().getDataOrganization());
+		}
+		
 		fireDataTypeManagerChanged(archive);
 	}
 
diff --git a/Ghidra/Framework/SoftwareModeling/src/main/java/ghidra/program/model/data/FileDataTypeManager.java b/Ghidra/Framework/SoftwareModeling/src/main/java/ghidra/program/model/data/FileDataTypeManager.java
index 80aa02855e3..097ef692864 100644
--- a/Ghidra/Framework/SoftwareModeling/src/main/java/ghidra/program/model/data/FileDataTypeManager.java
+++ b/Ghidra/Framework/SoftwareModeling/src/main/java/ghidra/program/model/data/FileDataTypeManager.java
@@ -192,6 +192,15 @@ public static File convertFilename(File file) {
 		fname = fname + SUFFIX;
 		return new File(file.getParentFile(), fname);
 	}
+	
+	/**
+	 * Set the data organization used by this data type manager.
+	 * 
+	 * @param org the new data organization
+	 */
+	public void setDataOrganization(DataOrganization org) {
+		dataOrganization = org;
+	}
 
 	private void updateRootCategoryName(ResourceFile newFile, Category root) {
 		String newName = getRootName(newFile.getName());
