diff --git a/Ghidra/Features/Base/src/main/java/ghidra/app/util/bin/format/pe/NTHeader.java b/Ghidra/Features/Base/src/main/java/ghidra/app/util/bin/format/pe/NTHeader.java
index 0912d754843..0d1a4145aea 100644
--- a/Ghidra/Features/Base/src/main/java/ghidra/app/util/bin/format/pe/NTHeader.java
+++ b/Ghidra/Features/Base/src/main/java/ghidra/app/util/bin/format/pe/NTHeader.java
@@ -154,7 +154,7 @@ public long rvaToPointer(long rva) {
 		SectionHeader[] sections = fileHeader.getSectionHeaders();
 		for (SectionHeader section : sections) {
 			long sectionVA = section.getVirtualAddress() & Conv.INT_MASK;
-			long rawSize = section.getSizeOfRawData() & Conv.INT_MASK;
+			long vSize = section.getVirtualSize() & Conv.INT_MASK;
 			long rawPtr = section.getPointerToRawData() & Conv.INT_MASK;
 
 			switch (layout) {
@@ -162,7 +162,7 @@ public long rvaToPointer(long rva) {
 					return rva;
 				case FILE:
 				default:
-					if (rva >= sectionVA && rva < sectionVA + rawSize) {
+					if (rva >= sectionVA && rva < sectionVA + vSize) {
 						return rva + rawPtr - sectionVA;
 					}
 					break;
