diff --git a/Ghidra/Features/Base/src/main/java/ghidra/app/script/GhidraScript.java b/Ghidra/Features/Base/src/main/java/ghidra/app/script/GhidraScript.java
index d1a295363da..5982fff77a7 100644
--- a/Ghidra/Features/Base/src/main/java/ghidra/app/script/GhidraScript.java
+++ b/Ghidra/Features/Base/src/main/java/ghidra/app/script/GhidraScript.java
@@ -1577,6 +1577,7 @@ public void setBackgroundColor(Address address, Color color) throws ImproperUseE
 				"The setBackgroundColor() method can only be used when running headed Ghidra.");
 		}
 
+		start();
 		PluginTool tool = state.getTool();
 		ColorizingService service = tool.getService(ColorizingService.class);
 		if (service == null) {
@@ -1586,6 +1587,7 @@ public void setBackgroundColor(Address address, Color color) throws ImproperUseE
 		}
 
 		service.setBackgroundColor(address, address, color);
+		end();
 	}
 
 	/**
