From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: astrelsky <ajs222@njit.edu>
Date: Sat, 1 Aug 2020 13:22:24 -0400
Subject: [PATCH] 2156: Added getDataTypeManager to GhidraScript

---
 .../java/ghidra/app/script/GhidraScript.java  | 25 +++++++++++++++++++
 1 file changed, 25 insertions(+)

diff --git a/Ghidra/Features/Base/src/main/java/ghidra/app/script/GhidraScript.java b/Ghidra/Features/Base/src/main/java/ghidra/app/script/GhidraScript.java
index 60e11eb7a1..f02283fa12 100644
--- a/Ghidra/Features/Base/src/main/java/ghidra/app/script/GhidraScript.java
+++ b/Ghidra/Features/Base/src/main/java/ghidra/app/script/GhidraScript.java
@@ -63,6 +63,7 @@ import ghidra.program.database.ProgramDB;
 import ghidra.program.flatapi.FlatProgramAPI;
 import ghidra.program.model.address.Address;
 import ghidra.program.model.address.AddressSetView;
+import ghidra.program.model.data.DataTypeManager;
 import ghidra.program.model.lang.*;
 import ghidra.program.model.listing.*;
 import ghidra.program.model.listing.CodeUnitFormatOptions.ShowBlockName;
@@ -4031,6 +4032,30 @@ public abstract class GhidraScript extends FlatProgramAPI {
 		return result;
 	}
 
+	/**
+	 * Gets the open DataTypeManager with the corresponding name.
+	 * <br><br>
+	 * NOTE: When using a DataTypeManager that is not the currentProgram's
+	 * it may be necessary to handle the transactions.
+	 * @param name the DataTypeManager's name
+	 * @return the DataTypeManager or null if not found
+	 * @see DataTypeManager#startTransaction(String)
+	 * @see DataTypeManager#endTransaction(int,boolean)
+	 */
+	public DataTypeManager getDataTypeManager(String name) {
+		DataTypeManagerService service = state.getTool().getService(DataTypeManagerService.class);
+		if (service == null) {
+			printerr("The DataTypeManagerService is not enabled");
+			return null;
+		}
+		for (DataTypeManager dtm : service.getDataTypeManagers()) {
+			if (dtm.getName().equals(name)) {
+				return dtm;
+			}
+		}
+		return null;
+	}
+
 	private class ScriptStatusListener implements StatusListener {
 		StringBuilder errors = new StringBuilder();
 
-- 
2.44.0

