diff --git a/Ghidra/Framework/SoftwareModeling/src/main/java/ghidra/program/model/pcode/PcodeBlock.java b/Ghidra/Framework/SoftwareModeling/src/main/java/ghidra/program/model/pcode/PcodeBlock.java
index 86798e982bc..17fd4cfbec9 100644
--- a/Ghidra/Framework/SoftwareModeling/src/main/java/ghidra/program/model/pcode/PcodeBlock.java
+++ b/Ghidra/Framework/SoftwareModeling/src/main/java/ghidra/program/model/pcode/PcodeBlock.java
@@ -91,7 +91,12 @@ public static String typeToName(int type) {
 
 	public static int nameToType(String name) {
 		switch (name.charAt(0)) {
+			case 'b':
+				return BASIC;
 			case 'c':
+				if (name.equals("condition")) {
+					return CONDITION;
+				}
 				return COPY;
 			case 'd':
 				return DOWHILE;
