From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: maruipu <maruipu@bytedance.com>
Date: Sat, 22 Aug 2020 22:17:39 +0800
Subject: [PATCH] 2221: [bugfix]fix duplicated imported symbol

---
 .../macho/commands/dyld/AbstractDyldInfoState.java     | 10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

diff --git a/Ghidra/Features/Base/src/main/java/ghidra/app/util/bin/format/macho/commands/dyld/AbstractDyldInfoState.java b/Ghidra/Features/Base/src/main/java/ghidra/app/util/bin/format/macho/commands/dyld/AbstractDyldInfoState.java
index d5c45eb74..488282a01 100644
--- a/Ghidra/Features/Base/src/main/java/ghidra/app/util/bin/format/macho/commands/dyld/AbstractDyldInfoState.java
+++ b/Ghidra/Features/Base/src/main/java/ghidra/app/util/bin/format/macho/commands/dyld/AbstractDyldInfoState.java
@@ -77,8 +77,14 @@ abstract public class AbstractDyldInfoState {
 
 	private Symbol getSymbol() {
 		SymbolIterator symbolIterator = program.getSymbolTable().getSymbols(symbolName);
-		if (symbolIterator.hasNext()) {
-			return symbolIterator.next();
+		while (symbolIterator.hasNext()) {
+			Symbol symbol = symbolIterator.next();
+			// In some situation, a imported symbol could appear both in `Binding Info` and `IndSym Table`.
+			// We select the one which is in the global name space.
+			if (symbol.getParentNamespace().isGlobal()
+				|| symbolIterator.hasNext() == false) {
+				return symbol;
+			}
 		}
 		return null;
 	}
-- 
2.39.0

