diff --git a/Ghidra/Features/Base/src/main/java/ghidra/app/util/bin/format/macho/commands/dyld/AbstractDyldInfoState.java b/Ghidra/Features/Base/src/main/java/ghidra/app/util/bin/format/macho/commands/dyld/AbstractDyldInfoState.java
index 633a7187795..bb9381337e3 100644
--- a/Ghidra/Features/Base/src/main/java/ghidra/app/util/bin/format/macho/commands/dyld/AbstractDyldInfoState.java
+++ b/Ghidra/Features/Base/src/main/java/ghidra/app/util/bin/format/macho/commands/dyld/AbstractDyldInfoState.java
@@ -77,8 +77,14 @@ final public void perform(TaskMonitor monitor) throws Exception {
 
 	private Symbol getSymbol() {
 		SymbolIterator symbolIterator = program.getSymbolTable().getSymbols(symbolName);
-		if (symbolIterator.hasNext()) {
-			return symbolIterator.next();
+		while (symbolIterator.hasNext()) {
+			Symbol symbol = symbolIterator.next();
+			// In some situation, a imported symbol could appear both in `Binding Info` and `IndSym Table`.
+			// We select the one which is in the global name space.
+			if (symbol.getParentNamespace().isGlobal()
+				|| symbolIterator.hasNext() == false) {
+				return symbol;
+			}
 		}
 		return null;
 	}
