diff --git a/.gitignore b/.gitignore
index 6cad771b5ca..a05aa486377 100644
--- a/.gitignore
+++ b/.gitignore
@@ -55,6 +55,9 @@ Release
 *.vcxproj.*
 .vs/
 
+# Ignore Visual Studio Code files
+.vscode/
+
 # Ignore UNIX backup files
 *~
 *.swp
diff --git a/gradle/support/ip.gradle b/gradle/support/ip.gradle
index 5e6c4a8ab3b..c17330d9082 100644
--- a/gradle/support/ip.gradle
+++ b/gradle/support/ip.gradle
@@ -119,6 +119,7 @@ def Map<String, List<String>> getIpForModule(Project p) {
 		exclude "**/data/build.xml" // language build file (generated for dev only)
 		exclude "**/.vs/**"
 		exclude "**/*.vcxproj.user"
+		exclude "**/.vscode/**"
 	}
 	tree.each { file ->
 			String ip = getIp(p.projectDir, file)
