diff --git a/Ghidra/Framework/SoftwareModeling/src/main/java/ghidra/program/database/data/DataTypeManagerDB.java b/Ghidra/Framework/SoftwareModeling/src/main/java/ghidra/program/database/data/DataTypeManagerDB.java
index ec17e24dbc2..e5e5dea03b4 100644
--- a/Ghidra/Framework/SoftwareModeling/src/main/java/ghidra/program/database/data/DataTypeManagerDB.java
+++ b/Ghidra/Framework/SoftwareModeling/src/main/java/ghidra/program/database/data/DataTypeManagerDB.java
@@ -683,14 +683,14 @@ public String getUniqueName(CategoryPath path1, CategoryPath path2, String baseN
 
 	@Override
 	public Category getCategory(CategoryPath path) {
-		if (path.equals(CategoryPath.ROOT)) {
-			return root;
-		}
-		Category parent = getCategory(path.getParent());
-		if (parent == null) {
-			return null;
+		Category cat = root;
+		for (String parent : path.asList()) {
+			cat = cat.getCategory(parent);
+			if (cat == null) {
+				return null;
+			}
 		}
-		return parent.getCategory(path.getName());
+		return cat;
 	}
 
 	CategoryDB getCategoryDB(long id) throws IOException {
