From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: astrelsky <ajs222@njit.edu>
Date: Mon, 9 Nov 2020 10:26:24 -0500
Subject: [PATCH] 2446: Non recursive DataTypeManagerDB.getCategory

---
 .../program/database/data/DataTypeManagerDB.java      | 11 +++++++----
 1 file changed, 7 insertions(+), 4 deletions(-)

diff --git a/Ghidra/Framework/SoftwareModeling/src/main/java/ghidra/program/database/data/DataTypeManagerDB.java b/Ghidra/Framework/SoftwareModeling/src/main/java/ghidra/program/database/data/DataTypeManagerDB.java
index 28cb21906c..b352416047 100644
--- a/Ghidra/Framework/SoftwareModeling/src/main/java/ghidra/program/database/data/DataTypeManagerDB.java
+++ b/Ghidra/Framework/SoftwareModeling/src/main/java/ghidra/program/database/data/DataTypeManagerDB.java
@@ -1174,11 +1174,14 @@ abstract public class DataTypeManagerDB implements DataTypeManager {
 		if (path.equals(CategoryPath.ROOT)) {
 			return root;
 		}
-		Category parent = getCategory(path.getParent());
-		if (parent == null) {
-			return null;
+		Category cat = root;
+		for (String parent : path.asList()) {
+			cat = cat.getCategory(parent);
+			if (cat == null) {
+				return null;
+			}
 		}
-		return parent.getCategory(path.getName());
+		return cat;
 	}
 
 	CategoryDB getCategoryDB(long id) throws IOException {
-- 
2.45.1

