diff --git a/Ghidra/Framework/SoftwareModeling/src/main/java/ghidra/program/model/data/DataTypeWriter.java b/Ghidra/Framework/SoftwareModeling/src/main/java/ghidra/program/model/data/DataTypeWriter.java
index 42088a817e6..1142f63f468 100644
--- a/Ghidra/Framework/SoftwareModeling/src/main/java/ghidra/program/model/data/DataTypeWriter.java
+++ b/Ghidra/Framework/SoftwareModeling/src/main/java/ghidra/program/model/data/DataTypeWriter.java
@@ -231,6 +231,9 @@ private void doWrite(DataType dt, TaskMonitor monitor, boolean throwExceptionOnI
 			return;
 		}
 		if (dt instanceof FunctionDefinition) {
+			writer.write(((FunctionDefinition) dt).getPrototypeString(true));
+			writer.write(";");
+			writer.write(EOL);
 			return;
 		}
 		if (dt instanceof FactoryDataType) {
