From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mikhail <neitri@gmail.com>
Date: Fri, 8 Jan 2021 20:00:04 +0600
Subject: [PATCH] 2633: __cdecl16far incorectly return 32bit value

---
 Ghidra/Processors/x86/data/languages/x86-16.cspec | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/Ghidra/Processors/x86/data/languages/x86-16.cspec b/Ghidra/Processors/x86/data/languages/x86-16.cspec
index 25019922b..9b7c3a602 100644
--- a/Ghidra/Processors/x86/data/languages/x86-16.cspec
+++ b/Ghidra/Processors/x86/data/languages/x86-16.cspec
@@ -121,6 +121,9 @@
       <pentry minsize="1" maxsize="2">
         <register name="AX"/>
       </pentry>
+      <pentry minsize="3" maxsize="4">
+        <addr space="join" piece1="DX" piece2="AX"/>
+      </pentry>
     </output>
     <unaffected>
       <register name="SP"/>
-- 
2.39.0

