diff --git a/Ghidra/Features/Base/src/main/java/ghidra/app/util/cparser/C/CParserUtils.java b/Ghidra/Features/Base/src/main/java/ghidra/app/util/cparser/C/CParserUtils.java
index 0354cad3e82..f54716478c2 100644
--- a/Ghidra/Features/Base/src/main/java/ghidra/app/util/cparser/C/CParserUtils.java
+++ b/Ghidra/Features/Base/src/main/java/ghidra/app/util/cparser/C/CParserUtils.java
@@ -23,6 +23,7 @@
 import ghidra.program.model.listing.Program;
 import ghidra.util.*;
 import ghidra.util.exception.DuplicateNameException;
+import org.apache.commons.lang3.StringUtils;
 
 public class CParserUtils {
 
@@ -165,6 +166,18 @@ public static FunctionDefinitionDataType parseSignature(DataTypeManagerService s
 			return null;
 		}
 
+        GenericCallingConvention convention = null;
+
+        for (GenericCallingConvention conv : GenericCallingConvention.values()) {
+            if (conv == GenericCallingConvention.unknown) continue;
+
+            if (StringUtils.endsWith(signatureParts[0], " " + conv)) {
+                convention = conv;
+                signatureParts[0] = StringUtils.removeEnd(signatureParts[0], " " + conv);
+                break;
+            }
+        }
+
 		String replacedText =
 			signatureParts[0] + " " + getTempName(signatureParts[1].length()) + signatureParts[2];
 
@@ -180,7 +193,9 @@ public static FunctionDefinitionDataType parseSignature(DataTypeManagerService s
 			// put back the old signature name
 			dt.setName(signatureParts[1]);
 
-			return (FunctionDefinitionDataType) dt;
+            FunctionDefinitionDataType fdt = (FunctionDefinitionDataType) dt;
+            if (convention != null) fdt.setGenericCallingConvention(convention);
+            return fdt;
 		}
 		catch (InvalidNameException | DuplicateNameException e) {
 			// can't happen since we are calling setName() with the value that was 
diff --git a/Ghidra/Features/Base/src/main/java/ghidra/app/util/xml/FunctionsXmlMgr.java b/Ghidra/Features/Base/src/main/java/ghidra/app/util/xml/FunctionsXmlMgr.java
index d5887d38694..d13e2644e37 100644
--- a/Ghidra/Features/Base/src/main/java/ghidra/app/util/xml/FunctionsXmlMgr.java
+++ b/Ghidra/Features/Base/src/main/java/ghidra/app/util/xml/FunctionsXmlMgr.java
@@ -155,6 +155,12 @@ void read(XmlPullParser parser, boolean overwriteConflicts, boolean ignoreStackF
 							Namespace namespace =
 								NamespaceUtils.getFunctionNamespaceAt(program, namespacePath,
 									entryPoint);
+							if(namespace == null && namespacePath != null){
+							    namespace = NamespaceUtils.createNamespaceHierarchy(
+							            namespacePath.getPath(),
+                                        program.getGlobalNamespace(),
+                                        program,SourceType.IMPORTED);
+                            }
 							if (namespace == null) {
 								namespace = program.getGlobalNamespace();
 							}
@@ -210,7 +216,11 @@ void read(XmlPullParser parser, boolean overwriteConflicts, boolean ignoreStackF
 						}
 					}
 					else {
-						tryToParseTypeInfoComment(monitor, func, typeInfoComment);
+                        tryToParseTypeInfoComment(monitor, func, typeInfoComment);
+
+                        if (func.getSignature().getGenericCallingConvention() == GenericCallingConvention.thiscall && func.getParentNamespace() != null) {
+                            NamespaceUtils.convertNamespaceToClass(func.getParentNamespace());
+                        }
 					}
 					addLocalVars(func, stackVariables, overwriteConflicts);
 
diff --git a/gradle/support/ip.gradle b/gradle/support/ip.gradle
index d8fc9f89d23..d984bc42c8c 100644
--- a/gradle/support/ip.gradle
+++ b/gradle/support/ip.gradle
@@ -133,6 +133,7 @@ def Map<String, List<String>> getIpForModule(Project p) {
 		exclude "**/data/build.xml" // language build file (generated for dev only)
 		exclude "**/.vs/**"
 		exclude "**/*.vcxproj.user"
+		exclude "**/*.iml" // Intellij project files
 	}
 	tree.each { file ->
 			String ip = getIp(p.projectDir, file)
