diff --git a/Ghidra/Features/Decompiler/src/main/java/ghidra/app/plugin/core/decompile/actions/RenameLocalAction.java b/Ghidra/Features/Decompiler/src/main/java/ghidra/app/plugin/core/decompile/actions/RenameLocalAction.java
index 3eb38c76508..cbb7cf6b2c6 100644
--- a/Ghidra/Features/Decompiler/src/main/java/ghidra/app/plugin/core/decompile/actions/RenameLocalAction.java
+++ b/Ghidra/Features/Decompiler/src/main/java/ghidra/app/plugin/core/decompile/actions/RenameLocalAction.java
@@ -71,6 +71,9 @@ protected boolean isEnabledForDecompilerContext(DecompilerActionContext context)
 		if (highSymbol == null) {
 			return false;
 		}
+		if (highSymbol instanceof EquateSymbol) {
+			return false;
+		}
 		return !highSymbol.isGlobal();
 	}
 
diff --git a/Ghidra/Features/Decompiler/src/main/java/ghidra/app/plugin/core/decompile/actions/RetypeLocalAction.java b/Ghidra/Features/Decompiler/src/main/java/ghidra/app/plugin/core/decompile/actions/RetypeLocalAction.java
index 03afbaec6fd..9ee71e39336 100644
--- a/Ghidra/Features/Decompiler/src/main/java/ghidra/app/plugin/core/decompile/actions/RetypeLocalAction.java
+++ b/Ghidra/Features/Decompiler/src/main/java/ghidra/app/plugin/core/decompile/actions/RetypeLocalAction.java
@@ -144,6 +144,9 @@ protected boolean isEnabledForDecompilerContext(DecompilerActionContext context)
 		if (highSymbol == null) {
 			return false;
 		}
+		if (highSymbol instanceof EquateSymbol) {
+			return false;
+		}
 		return !highSymbol.isGlobal();
 	}
 
