diff --git a/Ghidra/Features/Base/src/main/java/ghidra/app/util/opinion/ElfLoaderOptionsFactory.java b/Ghidra/Features/Base/src/main/java/ghidra/app/util/opinion/ElfLoaderOptionsFactory.java
index 9fe094e5a0c..3cd8ef0855a 100644
--- a/Ghidra/Features/Base/src/main/java/ghidra/app/util/opinion/ElfLoaderOptionsFactory.java
+++ b/Ghidra/Features/Base/src/main/java/ghidra/app/util/opinion/ElfLoaderOptionsFactory.java
@@ -182,6 +182,12 @@ private static String validateAddressSpaceOffsetOption(Option option, AddressSpa
 			return "Invalid type for option: " + name + " - " + option.getValueClass();
 		}
 		String value = (String) option.getValue();
+
+		/* Strip any leading 0x prefix */
+		if (value.startsWith("0x")) {
+			value = value.substring(2);
+		}
+
 		try {
 			space.getAddress(Long.parseUnsignedLong(value, 16), true);// verify valid address
 		}
