From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: WorksButNotTested <you@example.com>
Date: Fri, 25 Jun 2021 20:34:30 +0100
Subject: [PATCH] 3157: Accept leading 0x on Elf image base addresses

---
 .../ghidra/app/util/opinion/ElfLoaderOptionsFactory.java    | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/Ghidra/Features/Base/src/main/java/ghidra/app/util/opinion/ElfLoaderOptionsFactory.java b/Ghidra/Features/Base/src/main/java/ghidra/app/util/opinion/ElfLoaderOptionsFactory.java
index 70b1128fd..4faf3a040 100644
--- a/Ghidra/Features/Base/src/main/java/ghidra/app/util/opinion/ElfLoaderOptionsFactory.java
+++ b/Ghidra/Features/Base/src/main/java/ghidra/app/util/opinion/ElfLoaderOptionsFactory.java
@@ -213,6 +213,12 @@ public class ElfLoaderOptionsFactory {
 			return "Invalid type for option: " + name + " - " + option.getValueClass();
 		}
 		String value = (String) option.getValue();
+
+		/* Strip any leading 0x prefix */
+		if (value.startsWith("0x")) {
+			value = value.substring(2);
+		}
+
 		try {
 			space.getAddress(Long.parseUnsignedLong(value, 16), true);// verify valid address
 		}
-- 
2.43.0

