diff --git a/Ghidra/Features/Decompiler/src/decompile/cpp/slgh_compile.cc b/Ghidra/Features/Decompiler/src/decompile/cpp/slgh_compile.cc
index 45ccc9225b4..f9aea157c13 100644
--- a/Ghidra/Features/Decompiler/src/decompile/cpp/slgh_compile.cc
+++ b/Ghidra/Features/Decompiler/src/decompile/cpp/slgh_compile.cc
@@ -1229,7 +1229,7 @@ void ConsistencyChecker::examineVn(map<uintb,OptimizeRecord> &recs,
   if (vn->getOffset().getType() != ConstTpl::real) return;
 
   map<uintb,OptimizeRecord>::iterator iter;
-  iter = recs.insert( pair<uint4,OptimizeRecord>(vn->getOffset().getReal(),OptimizeRecord())).first;
+  iter = recs.insert({vn->getOffset().getReal(),OptimizeRecord()}).first;
   if (inslot>=0) {
     (*iter).second.readop = i;
     (*iter).second.readcount += 1;
