From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Zach Frichtl <ZFrichtl@gmail.com>
Date: Mon, 25 Oct 2021 07:11:04 -0400
Subject: [PATCH] 3546: Add support for R_AARCH64_MOVW_UABS_G* relocations.

---
 .../AARCH64_ElfRelocationHandler.java          | 18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

diff --git a/Ghidra/Processors/AARCH64/src/main/java/ghidra/app/util/bin/format/elf/relocation/AARCH64_ElfRelocationHandler.java b/Ghidra/Processors/AARCH64/src/main/java/ghidra/app/util/bin/format/elf/relocation/AARCH64_ElfRelocationHandler.java
index aff6978ab..754956bb2 100644
--- a/Ghidra/Processors/AARCH64/src/main/java/ghidra/app/util/bin/format/elf/relocation/AARCH64_ElfRelocationHandler.java
+++ b/Ghidra/Processors/AARCH64/src/main/java/ghidra/app/util/bin/format/elf/relocation/AARCH64_ElfRelocationHandler.java
@@ -128,6 +128,24 @@ public class AARCH64_ElfRelocationHandler extends ElfRelocationHandler {
 				break;
 			}
 
+			// MOV[ZK]:   ((S+A) >>  shift) & 0xffff
+			case AARCH64_ElfRelocationConstants.R_AARCH64_MOVW_UABS_G0:
+			case AARCH64_ElfRelocationConstants.R_AARCH64_MOVW_UABS_G0_NC:
+			case AARCH64_ElfRelocationConstants.R_AARCH64_MOVW_UABS_G1:
+			case AARCH64_ElfRelocationConstants.R_AARCH64_MOVW_UABS_G1_NC:
+			case AARCH64_ElfRelocationConstants.R_AARCH64_MOVW_UABS_G2:
+			case AARCH64_ElfRelocationConstants.R_AARCH64_MOVW_UABS_G2_NC:
+			case AARCH64_ElfRelocationConstants.R_AARCH64_MOVW_UABS_G3: {
+				int oldValue = memory.getInt(relocationAddress, isBigEndianInstructions);
+				long result = symbolValue + addend;
+				int shift = (oldValue >> 21) & 0x3;  // Shift amount is in bits 22:21
+
+				newValue = (oldValue & 0xffe0001f) | (((result >> shift * 16) & 0xffff) << 5);
+
+				memory.setInt(relocationAddress, (int) newValue, isBigEndianInstructions);
+				break;
+			}
+
 			// ADRH: ((PG(S+A)-PG(P)) >> 12) & 0x1fffff
 			case AARCH64_ElfRelocationConstants.R_AARCH64_ADR_PREL_PG_HI21:
 			case AARCH64_ElfRelocationConstants.R_AARCH64_P32_ADR_PREL_PG_HI21: {
-- 
2.39.0

