diff --git a/Ghidra/Features/Decompiler/src/decompile/cpp/sleigh_arch.cc b/Ghidra/Features/Decompiler/src/decompile/cpp/sleigh_arch.cc
index a8efac8c7ea..4046c48cac5 100644
--- a/Ghidra/Features/Decompiler/src/decompile/cpp/sleigh_arch.cc
+++ b/Ghidra/Features/Decompiler/src/decompile/cpp/sleigh_arch.cc
@@ -464,5 +464,5 @@ void SleighArchitecture::shutdown(void)
   for(map<int4,Sleigh *>::const_iterator iter=translators.begin();iter!=translators.end();++iter)
     delete (*iter).second;
   translators.clear();
-  // description.clear();  // static vector is destroyed by the normal exit handler
+  description.clear();  // static vector is destroyed by the normal exit handler
 }
