From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Alessandro Gatti <a.gatti@frob.it>
Date: Thu, 9 Dec 2021 19:56:14 +0100
Subject: [PATCH] 3734: More stringent size checks for Sleigh spaces.

---
 .../ghidra/sleigh/grammar/SleighCompiler.g    | 19 +++++++++++++++----
 1 file changed, 15 insertions(+), 4 deletions(-)

diff --git a/Ghidra/Framework/SoftwareModeling/src/main/antlr/ghidra/sleigh/grammar/SleighCompiler.g b/Ghidra/Framework/SoftwareModeling/src/main/antlr/ghidra/sleigh/grammar/SleighCompiler.g
index f815601a4..cb8158b94 100644
--- a/Ghidra/Framework/SoftwareModeling/src/main/antlr/ghidra/sleigh/grammar/SleighCompiler.g
+++ b/Ghidra/Framework/SoftwareModeling/src/main/antlr/ghidra/sleigh/grammar/SleighCompiler.g
@@ -398,9 +398,11 @@ contextdef
 spacedef
 	scope {
 		SpaceQuality quality;
+		boolean wordSizeSet;
 	}
 	@init {
 		$spacedef::quality = null;
+		$spacedef::wordSizeSet = false;
 	}
 	:	^(OP_SPACE n=unbound_identifier["space"] {
 			String name = "<parse error>";
@@ -441,14 +443,23 @@ typemod
 	;
 
 sizemod
-	:	^(OP_SIZE i=integer) {
-			$spacedef::quality.size = $i.value.intValue();
+	:	^(t=OP_SIZE i=integer) {
+			if ($spacedef::quality.size != 0) {
+				reportError(find(t), "space '" + $spacedef::quality.name + "' already has a size set");
+			} else {
+				$spacedef::quality.size = $i.value.intValue();
+			}
 		}
 	;
 
 wordsizemod
-	:	^(OP_WORDSIZE i=integer) {
-			$spacedef::quality.wordsize = $i.value.intValue();
+	:	^(t=OP_WORDSIZE i=integer) {
+			if ($spacedef::wordSizeSet) {
+				reportError(find(t), "space '" + $spacedef::quality.name + "' already has a word size set");
+			} else {
+				$spacedef::quality.wordsize = $i.value.intValue();
+				$spacedef::wordSizeSet = true;
+			}
 		}
 	;
 
-- 
2.41.0

