diff --git a/Ghidra/Features/PDB/src/main/java/ghidra/app/util/bin/format/pdb/ApplyFunctions.java b/Ghidra/Features/PDB/src/main/java/ghidra/app/util/bin/format/pdb/ApplyFunctions.java
index 5a4f3975b8e..1573e361733 100644
--- a/Ghidra/Features/PDB/src/main/java/ghidra/app/util/bin/format/pdb/ApplyFunctions.java
+++ b/Ghidra/Features/PDB/src/main/java/ghidra/app/util/bin/format/pdb/ApplyFunctions.java
@@ -112,7 +112,7 @@ private static Function checkInsideThunkFunction(Program program, Address addres
 			return null;
 		}
 		AddressSet thunkBody = new AddressSet(thunkFunction.getBody());
-		boolean hasRangeMatch = true;
+		boolean hasRangeMatch = false;
 		AddressRangeIterator ari = thunkBody.getAddressRanges(true);
 		while (ari.hasNext()) {
 			if (monitor.isCancelled()) {
