diff --git a/Ghidra/Features/Base/src/main/java/ghidra/program/flatapi/FlatProgramAPI.java b/Ghidra/Features/Base/src/main/java/ghidra/program/flatapi/FlatProgramAPI.java
index 33933b397e5..6bc3eed4425 100644
--- a/Ghidra/Features/Base/src/main/java/ghidra/program/flatapi/FlatProgramAPI.java
+++ b/Ghidra/Features/Base/src/main/java/ghidra/program/flatapi/FlatProgramAPI.java
@@ -1902,6 +1902,32 @@ public final byte[] getBytes(Address address, int length) throws MemoryAccessExc
 		return bytes;
 	}
 
+	/**
+	 * Reads length number of bytes starting at the specified address as they appear in
+	 * the original binary.
+	 * Note: this could be inefficient if length is large
+	 * @param address the address to start reading
+	 * @param length the number of bytes to read
+	 * @return an array of bytes
+	 * @see ghidra.program.model.mem.Memory
+	 */
+	protected byte[] getOriginalBytes(Address address, int length) {
+		MemoryBlock stringMemoryBlock = currentProgram.getMemory().getBlock(address);
+		if (stringMemoryBlock == null)
+			return null;
+		FileBytes fileBytes = currentProgram.getMemory().getAllFileBytes().get(0);
+		MemoryBlockSourceInfo memoryInformation = stringMemoryBlock.getSourceInfos().get(0);
+		long fileOffset = address.getOffset() - memoryInformation.getMinAddress().getOffset()
+				+ memoryInformation.getFileBytesOffset();
+		try {
+			byte[] result = new byte[length];
+			fileBytes.getOriginalBytes(fileOffset, result);
+			return result;
+		} catch (IOException X) {
+			return null;
+		}
+	}
+
 	/**
 	 * Sets the 'byte' value at the specified address.
 	 * @param address the address to set the 'byte'
