diff --git a/Ghidra/Processors/68000/data/languages/68000.sinc b/Ghidra/Processors/68000/data/languages/68000.sinc
index 3a7dbb4a45b..241116a32a7 100644
--- a/Ghidra/Processors/68000/data/languages/68000.sinc
+++ b/Ghidra/Processors/68000/data/languages/68000.sinc
@@ -1720,7 +1720,7 @@ ptestLevel: "#"^mregn	is mregn  { export *[const]:1 mregn; }
 :rts				is opbig=0x4e & op37=14 & op02=5				{ PC = *SP; SP = SP+4; return [PC]; }
 
 :sbcd Tyb,Txb			is op=8 & op48=16 & Txb & Tyb
-                    {  CF = (Tyb < Txb) || ( (XF == 1) && (Tyb == Txb) );  Tyb = Tyb - Txb - XF; Tyb = bcdAdjust(Tyb); bcdflags(Tyb); }
+                    {  CF = (Txb < Tyb) || ( (XF == 1) && (Txb == Tyb) );  Txb = Txb - Tyb - XF; Txb = bcdAdjust(Txb); bcdflags(Txb); }
 
 :s^cc eab			is (op=5 & cc & op67=3 & $(DAT_ALTER_ADDR_MODES))... & eab				{ eab = -cc; }
 
