diff --git a/Ghidra/Features/GnuDemangler/src/main/java/ghidra/app/util/demangler/gnu/GnuDemangler.java b/Ghidra/Features/GnuDemangler/src/main/java/ghidra/app/util/demangler/gnu/GnuDemangler.java
index 3b1735927af..3941f386c12 100644
--- a/Ghidra/Features/GnuDemangler/src/main/java/ghidra/app/util/demangler/gnu/GnuDemangler.java
+++ b/Ghidra/Features/GnuDemangler/src/main/java/ghidra/app/util/demangler/gnu/GnuDemangler.java
@@ -56,6 +56,11 @@ public boolean canDemangle(Program program) {
 		if (!specId.toLowerCase().contains("windows")) {
 			return true;
 		}
+
+		String compiler = program.getCompiler();
+		if (compiler.startsWith("gcc:")) {
+			return true;
+		}
 		return false;
 	}
 
