diff --git a/Ghidra/Features/Python/ghidra_scripts/ImportSymbolsScript.py b/Ghidra/Features/Python/ghidra_scripts/ImportSymbolsScript.py
index 20802dcbeb8..fb127d1d649 100644
--- a/Ghidra/Features/Python/ghidra_scripts/ImportSymbolsScript.py
+++ b/Ghidra/Features/Python/ghidra_scripts/ImportSymbolsScript.py
@@ -33,9 +33,17 @@
 
 f = askFile("Give me a file to open", "Go baby go!")
 
-for line in file(f.absolutePath):  # note, cannot use open(), since that is in GhidraScript
+for line_no, line in enumerate(file(f.absolutePath)):  # note, cannot use open(), since that is in GhidraScript
     pieces = line.split()
 
+    # skip empty lines and comments
+    if not pieces or line[0].startswith('#'):
+        continue
+   
+    if len(pieces) not in [2,3]:
+        print("Skip line #{}: {}".format(line_no, line))
+        continue
+
     name = pieces[0]
     address = toAddr(pieces[1])
 
