diff --git a/Ghidra/Features/Decompiler/src/main/doc/commonprofile.xsl b/Ghidra/Features/Decompiler/src/main/doc/commonprofile.xsl
index 6e79c8417f4..06ab4a58b85 100644
--- a/Ghidra/Features/Decompiler/src/main/doc/commonprofile.xsl
+++ b/Ghidra/Features/Decompiler/src/main/doc/commonprofile.xsl
@@ -2,5 +2,5 @@
 <xsl:stylesheet
     xmlns:xsl="http://www.w3.org/1999/XSL/Transform" version="1.0">
 
-  <xsl:import href="http://docbook.sourceforge.net/release/xsl/current/profiling/profile.xsl"/>
+  <xsl:import href="http://docbook.sourceforge.net/release/xsl/1.79.1/profiling/profile.xsl"/>
 </xsl:stylesheet>
diff --git a/Ghidra/Features/Decompiler/src/main/doc/cspec_html.xsl b/Ghidra/Features/Decompiler/src/main/doc/cspec_html.xsl
index 978d2a6b08f..7134ea93464 100644
--- a/Ghidra/Features/Decompiler/src/main/doc/cspec_html.xsl
+++ b/Ghidra/Features/Decompiler/src/main/doc/cspec_html.xsl
@@ -2,7 +2,7 @@
 <xsl:stylesheet
     xmlns:xsl="http://www.w3.org/1999/XSL/Transform" version="1.0">
 
-<xsl:import href="http://docbook.sourceforge.net/release/xsl/current/html/chunk.xsl"/>
+<xsl:import href="http://docbook.sourceforge.net/release/xsl/1.79.1/html/chunk.xsl"/>
 
 <xsl:param name="generate.toc">
   article/appendix  nop
diff --git a/Ghidra/Features/Decompiler/src/main/doc/decompileplugin_html.xsl b/Ghidra/Features/Decompiler/src/main/doc/decompileplugin_html.xsl
index f1744f6669b..9a97aa10109 100644
--- a/Ghidra/Features/Decompiler/src/main/doc/decompileplugin_html.xsl
+++ b/Ghidra/Features/Decompiler/src/main/doc/decompileplugin_html.xsl
@@ -2,7 +2,7 @@
 <xsl:stylesheet
     xmlns:xsl="http://www.w3.org/1999/XSL/Transform" version="1.0">
 
-<xsl:import href="http://docbook.sourceforge.net/release/xsl/current/html/chunk.xsl"/>
+<xsl:import href="http://docbook.sourceforge.net/release/xsl/1.79.1/html/chunk.xsl"/>
 
 <xsl:include href="decompileplugin_common.xsl" />
 
diff --git a/Ghidra/Features/Decompiler/src/main/doc/decompileplugin_pdf.xsl b/Ghidra/Features/Decompiler/src/main/doc/decompileplugin_pdf.xsl
index e315726e5c7..a8efbe331ee 100644
--- a/Ghidra/Features/Decompiler/src/main/doc/decompileplugin_pdf.xsl
+++ b/Ghidra/Features/Decompiler/src/main/doc/decompileplugin_pdf.xsl
@@ -2,7 +2,7 @@
 <xsl:stylesheet
     xmlns:xsl="http://www.w3.org/1999/XSL/Transform" version="1.0">
 
-<xsl:import href="http://docbook.sourceforge.net/release/xsl/current/fo/docbook.xsl"/>
+<xsl:import href="http://docbook.sourceforge.net/release/xsl/1.79.1/fo/docbook.xsl"/>
 
 <xsl:include href="decompileplugin_common.xsl" />
 
diff --git a/Ghidra/Features/Decompiler/src/main/doc/main_html.xsl b/Ghidra/Features/Decompiler/src/main/doc/main_html.xsl
index 802a9939a66..87c578f04bc 100644
--- a/Ghidra/Features/Decompiler/src/main/doc/main_html.xsl
+++ b/Ghidra/Features/Decompiler/src/main/doc/main_html.xsl
@@ -2,7 +2,7 @@
 <xsl:stylesheet
     xmlns:xsl="http://www.w3.org/1999/XSL/Transform" version="1.0">
 
-<xsl:import href="http://docbook.sourceforge.net/release/xsl/current/html/docbook.xsl"/>
+<xsl:import href="http://docbook.sourceforge.net/release/xsl/1.79.1/html/docbook.xsl"/>
 
 <xsl:param name="generate.toc">
   article/appendix  nop
diff --git a/Ghidra/Features/Decompiler/src/main/doc/pcoderef_html.xsl b/Ghidra/Features/Decompiler/src/main/doc/pcoderef_html.xsl
index 931e36975e4..b1d410e1f59 100644
--- a/Ghidra/Features/Decompiler/src/main/doc/pcoderef_html.xsl
+++ b/Ghidra/Features/Decompiler/src/main/doc/pcoderef_html.xsl
@@ -2,7 +2,7 @@
 <xsl:stylesheet
     xmlns:xsl="http://www.w3.org/1999/XSL/Transform" version="1.0">
 
-<xsl:import href="http://docbook.sourceforge.net/release/xsl/current/html/chunk.xsl"/>
+<xsl:import href="http://docbook.sourceforge.net/release/xsl/1.79.1/html/chunk.xsl"/>
 
 <xsl:include href="pcoderef_common.xsl" />
 
diff --git a/Ghidra/Features/Decompiler/src/main/doc/pcoderef_pdf.xsl b/Ghidra/Features/Decompiler/src/main/doc/pcoderef_pdf.xsl
index 0879af176b4..5f52f8017b1 100644
--- a/Ghidra/Features/Decompiler/src/main/doc/pcoderef_pdf.xsl
+++ b/Ghidra/Features/Decompiler/src/main/doc/pcoderef_pdf.xsl
@@ -2,7 +2,7 @@
 <xsl:stylesheet
     xmlns:xsl="http://www.w3.org/1999/XSL/Transform" version="1.0">
 
-<xsl:import href="http://docbook.sourceforge.net/release/xsl/current/fo/docbook.xsl"/>
+<xsl:import href="http://docbook.sourceforge.net/release/xsl/1.79.1/fo/docbook.xsl"/>
 
 <xsl:template match="table" mode="label.markup"/>
 
diff --git a/Ghidra/Features/Decompiler/src/main/doc/sleigh_html.xsl b/Ghidra/Features/Decompiler/src/main/doc/sleigh_html.xsl
index 15f1f655cf9..c16d801ec6f 100644
--- a/Ghidra/Features/Decompiler/src/main/doc/sleigh_html.xsl
+++ b/Ghidra/Features/Decompiler/src/main/doc/sleigh_html.xsl
@@ -2,7 +2,7 @@
 <xsl:stylesheet
     xmlns:xsl="http://www.w3.org/1999/XSL/Transform" version="1.0">
 
-<xsl:import href="http://docbook.sourceforge.net/release/xsl/current/html/chunk.xsl"/>
+<xsl:import href="http://docbook.sourceforge.net/release/xsl/1.79.1/html/chunk.xsl"/>
 
 <xsl:include href="sleigh_common.xsl" />
 
diff --git a/Ghidra/Features/Decompiler/src/main/doc/sleigh_pdf.xsl b/Ghidra/Features/Decompiler/src/main/doc/sleigh_pdf.xsl
index ae015f403d7..26420041737 100644
--- a/Ghidra/Features/Decompiler/src/main/doc/sleigh_pdf.xsl
+++ b/Ghidra/Features/Decompiler/src/main/doc/sleigh_pdf.xsl
@@ -2,7 +2,7 @@
 <xsl:stylesheet
     xmlns:xsl="http://www.w3.org/1999/XSL/Transform" version="1.0">
 
-<xsl:import href="http://docbook.sourceforge.net/release/xsl/current/fo/docbook.xsl"/>
+<xsl:import href="http://docbook.sourceforge.net/release/xsl/1.79.1/fo/docbook.xsl"/>
 
 <xsl:include href="sleigh_common.xsl" />
 
