From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: isJuhn <johanhellgren.98@gmail.com>
Date: Tue, 11 Oct 2022 19:37:15 +0200
Subject: [PATCH] 4653: Stub Cell-specific altivec instructions

---
 .../Processors/PowerPC/data/languages/altivec.sinc   | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/Ghidra/Processors/PowerPC/data/languages/altivec.sinc b/Ghidra/Processors/PowerPC/data/languages/altivec.sinc
index 7f4b339b4..88ca43ee2 100644
--- a/Ghidra/Processors/PowerPC/data/languages/altivec.sinc
+++ b/Ghidra/Processors/PowerPC/data/languages/altivec.sinc
@@ -12,6 +12,8 @@ define pcodeop loadVectorForShiftLeft;
 define pcodeop loadVectorForShiftRight;
 define pcodeop loadVectorIndexed;
 define pcodeop loadVectorIndexedLRU;
+define pcodeop loadVectorLeftIndexed;
+define pcodeop loadVectorRightIndexed;
 define pcodeop moveFromVectorStatusAndControlRegister;
 define pcodeop moveToVectorStatusAndControlRegister;
 define pcodeop storeVectorElementByteIndexed;
@@ -246,6 +248,16 @@ define pcodeop vectorUnpackLowSignedHalfWord;
 	vrD = loadVectorIndexedLRU(A,B);
 }
 
+:lvlx vrD,RA_OR_ZERO,B      is OP=31 & vrD & RA_OR_ZERO & B & XOP_1_10=519 & Rc=0
+{
+    vrD = loadVectorLeftIndexed(RA_OR_ZERO,B);
+}
+
+:lvrx vrD,RA_OR_ZERO,B      is OP=31 & vrD & RA_OR_ZERO & B & XOP_1_10=551 & Rc=0
+{
+    vrD = loadVectorRightIndexed(RA_OR_ZERO,B);
+}
+
 :mfvscr vrD			is OP=4 & vrD & vrAR=0 & vrBR=0 & XOP_1_10=770 & Rc=0
 {   # TODO definition
 	vrD = moveFromVectorStatusAndControlRegister();
-- 
2.40.1

