diff --git a/Ghidra/Features/Decompiler/src/decompile/cpp/slghsymbol.hh b/Ghidra/Features/Decompiler/src/decompile/cpp/slghsymbol.hh
index 82d95ffeff4..9dd5894465a 100644
--- a/Ghidra/Features/Decompiler/src/decompile/cpp/slghsymbol.hh
+++ b/Ghidra/Features/Decompiler/src/decompile/cpp/slghsymbol.hh
@@ -88,7 +88,7 @@ public:
   SleighSymbol *findSymbol(const string &nm) const { return findSymbolInternal(curscope,nm); }
   SleighSymbol *findSymbol(const string &nm,int4 skip) const { return findSymbolInternal(skipScope(skip),nm); }
   SleighSymbol *findGlobalSymbol(const string &nm) const { return findSymbolInternal(table[0],nm); }
-  SleighSymbol *findSymbol(uintm id) const { return symbollist[id]; }
+  SleighSymbol *findSymbol(uintm id) const { if (id < symbollist.size()) return symbollist[id]; else return NULL; }
   void replaceSymbol(SleighSymbol *a,SleighSymbol *b);
   void saveXml(ostream &s) const;
   void restoreXml(const Element *el,SleighBase *trans);
