diff --git a/Ghidra/Features/Decompiler/ghidra_scripts/SwitchOverride.java b/Ghidra/Features/Decompiler/ghidra_scripts/SwitchOverride.java
index 68d35baa88c..bb29b513c47 100644
--- a/Ghidra/Features/Decompiler/ghidra_scripts/SwitchOverride.java
+++ b/Ghidra/Features/Decompiler/ghidra_scripts/SwitchOverride.java
@@ -83,7 +83,7 @@ private boolean isComputedBranchInstruction(Instruction instr) {
 		
 		FlowType flowType = instr.getFlowType();
 		
-		if (flowType == RefType.COMPUTED_JUMP) {
+		if (flowType.isJump() && flowType.isComputed()) {
 			return true;
 		}
 		if (flowType.isCall()) {
