From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Phil Pemberton <philpem@philpem.me.uk>
Date: Wed, 8 Feb 2023 10:04:00 +0000
Subject: [PATCH] 4971: 65C02: Fix TRB/TSB instructions, fixes #4921

---
 Ghidra/Processors/6502/data/languages/65c02.slaspec | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/Ghidra/Processors/6502/data/languages/65c02.slaspec b/Ghidra/Processors/6502/data/languages/65c02.slaspec
index d936493a0..39befd712 100644
--- a/Ghidra/Processors/6502/data/languages/65c02.slaspec
+++ b/Ghidra/Processors/6502/data/languages/65c02.slaspec
@@ -208,7 +208,8 @@ ADDRIX: (imm16,X) is X; imm16     { addr:2 = imm16 + zext(X); tmp:2 = *:2 addr;
 :TRB OPTB                      is (tcc=0 & taaa=0 & td=1) ... & OPTB
 {
     local op1 = OPTB;
-    local result = A & ~op1;
+    local result = op1 & ~A;
+    OPTB = result;
     Z = result == 0;
 }
 
@@ -216,5 +217,6 @@ ADDRIX: (imm16,X) is X; imm16     { addr:2 = imm16 + zext(X); tmp:2 = *:2 addr;
 {
     local op1 = OPTB;
     local result = A | op1;
+    OPTB = result;
     Z = result == 0;
 }
-- 
2.39.1

