From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Andrew Strelsky <ajs222@njit.edu>
Date: Sat, 25 Feb 2023 08:48:28 -0500
Subject: [PATCH] 5032: Added missing LoadException constructors

---
 .../app/util/opinion/LoadException.java       | 19 +++++++++++++++++++
 1 file changed, 19 insertions(+)

diff --git a/Ghidra/Features/Base/src/main/java/ghidra/app/util/opinion/LoadException.java b/Ghidra/Features/Base/src/main/java/ghidra/app/util/opinion/LoadException.java
index 1c4cf83cb..54b7f53d5 100644
--- a/Ghidra/Features/Base/src/main/java/ghidra/app/util/opinion/LoadException.java
+++ b/Ghidra/Features/Base/src/main/java/ghidra/app/util/opinion/LoadException.java
@@ -37,4 +37,23 @@ public class LoadException extends IOException {
 	public LoadException(String message) {
 		super(message);
 	}
+
+	/**
+	 * Create a new {@link LoadException} with the given message and cause
+	 *
+	 * @param message The exception message
+	 * @param cause The exception cause
+	 */
+	public LoadException(String message, Throwable cause) {
+		super(message, cause);
+	}
+
+	/**
+	 * Create a new {@link LoadException} with the given cause
+	 *
+	 * @param cause The exception cause
+	 */
+	public LoadException(Throwable cause) {
+		super(cause);
+	}
 }
-- 
2.39.1

