From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: zebrapurring <>
Date: Sun, 30 Apr 2023 22:26:14 +0200
Subject: [PATCH] 5277: Fix srvAdmin compatibility with BSD systems

---
 Ghidra/RuntimeScripts/Linux/server/svrAdmin | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Ghidra/RuntimeScripts/Linux/server/svrAdmin b/Ghidra/RuntimeScripts/Linux/server/svrAdmin
index 41ebb345b..8ef590603 100755
--- a/Ghidra/RuntimeScripts/Linux/server/svrAdmin
+++ b/Ghidra/RuntimeScripts/Linux/server/svrAdmin
@@ -42,7 +42,7 @@ else
 fi
 
 # Identify server process owner if set within server.conf
-OWNER="$(grep '^wrapper.app.account=' "${CONFIG}" | sed -e 's/^.*=\(.*\)\s*.*$/\1/')"
+OWNER="$(grep '^wrapper.app.account=' "${CONFIG}" | cut -d '=' -f 2)"
 
 if [ -z "${OWNER}" -o "${OWNER}" = "$(whoami)" ]; then
 	VMARGS="-DUserAdmin.invocation=$(basename "${SCRIPT_FILE}")"
-- 
2.41.0

