From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Klaus=20K=C3=A4mpf?= <kkaempf@gmail.com>
Date: Thu, 11 May 2023 21:16:41 +0200
Subject: [PATCH] 5322: exclude generated sleigh files from IP scan
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Signed-off-by: Klaus Kämpf <kkaempf@gmail.com>
---
 gradle/support/ip.gradle | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/gradle/support/ip.gradle b/gradle/support/ip.gradle
index 65fdeb6c2..6be24c9f4 100644
--- a/gradle/support/ip.gradle
+++ b/gradle/support/ip.gradle
@@ -146,6 +146,17 @@ def Map<String, List<String>> getIpForModule(Project p) {
 		exclude "**/*.vcxproj.user"
 		exclude "**/.vscode/**"
 		exclude "**/*.iml" // Intellij project files
+                exclude "**/pcodeparse.cc"
+                exclude "**/slghparse.cc"
+                exclude "**/slghparse.hh"
+                exclude "**/slghscan.cc"
+                exclude "**/xml.cc"
+                exclude "**/sla_dbg/**"
+                exclude "**/sla_opt/**"
+                exclude "**/com_dbg/**"
+                exclude "**/com_opt/**"
+                exclude "**/sleigh_opt"
+                exclude "**/sleigh_dbg"
 	}
 	tree.each { file ->
 			String ip = getIp(p.projectDir, file)
-- 
2.40.0

