From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Andrew Strelsky <ajs222@njit.edu>
Date: Sat, 27 Apr 2024 05:44:09 -0400
Subject: [PATCH] 6462: Remove c keywords when parsing function signatures

---
 .../ghidra/app/util/parser/FunctionSignatureParser.java  | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/Ghidra/Features/Base/src/main/java/ghidra/app/util/parser/FunctionSignatureParser.java b/Ghidra/Features/Base/src/main/java/ghidra/app/util/parser/FunctionSignatureParser.java
index fc311a53bf..dec4567244 100644
--- a/Ghidra/Features/Base/src/main/java/ghidra/app/util/parser/FunctionSignatureParser.java
+++ b/Ghidra/Features/Base/src/main/java/ghidra/app/util/parser/FunctionSignatureParser.java
@@ -161,6 +161,9 @@ public class FunctionSignatureParser {
 			throw new ParseException("Can't parse function arguments");
 		}
 		String trailingText = newSignatureText.substring(endIndex + 1);
+
+		// remove any trailing whitespace and semicolon
+		trailingText = trailingText.replaceAll("\\s*;?\\s*$", "");
 		if (trailingText.trim().length() > 0) {
 			throw new ParseException(
 				"Unexpected trailing text at end of function: " + trailingText);
@@ -192,6 +195,9 @@ public class FunctionSignatureParser {
 			throw new ParseException("Missing parameter");
 		}
 
+		// remove c keywords that Ghidra doesn't support
+		arg = arg.replaceAll("\\b((const)|(volatile)|(restrict)|(struct)|(union))\\b", "");
+
 		// Attempt to resolve parameter assuming only a datatype is specified
 		DataType dt = resolveDataType(arg);
 		if (dt != null) {
@@ -261,6 +267,9 @@ public class FunctionSignatureParser {
 		}
 		String returnTypeName = StringUtils.join(split, " ", 0, split.length - 1);
 
+		// remove any c keywords from the return type name
+		returnTypeName = returnTypeName.replaceAll("\\b((static)|(inline)|(const)|(volatile)|(restrict)|(struct)|(union))\\b", "");
+
 		DataType dt = resolveDataType(returnTypeName);
 		if (dt == null) {
 			throw new ParseException("Can't resolve return type: " + returnTypeName);
-- 
2.45.1

