From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HEKAR Research Institute <contact@hekar.org>
Date: Sun, 17 Aug 2025 11:25:40 +0200
Subject: [PATCH] 8443: [fix] add support for little endian `VariableStorage`

---
 .../program/model/listing/VariableStorage.java     | 14 ++++++++++++--
 1 file changed, 12 insertions(+), 2 deletions(-)

diff --git a/Ghidra/Framework/SoftwareModeling/src/main/java/ghidra/program/model/listing/VariableStorage.java b/Ghidra/Framework/SoftwareModeling/src/main/java/ghidra/program/model/listing/VariableStorage.java
index ffcfe9cb68..9f8cbb17db 100644
--- a/Ghidra/Framework/SoftwareModeling/src/main/java/ghidra/program/model/listing/VariableStorage.java
+++ b/Ghidra/Framework/SoftwareModeling/src/main/java/ghidra/program/model/listing/VariableStorage.java
@@ -431,14 +431,24 @@ public class VariableStorage implements Comparable<VariableStorage> {
 	 * @return first varnode within the ordered list of varnodes
 	 */
 	public Varnode getFirstVarnode() {
-		return (varnodes == null || varnodes.length == 0) ? null : varnodes[0];
+		Language language = programArch == null ? null : programArch.getLanguage();
+		if (language != null && !language.isBigEndian()) {
+			return (varnodes == null || varnodes.length == 0) ? null : varnodes[varnodes.length - 1];
+		} else {
+			return (varnodes == null || varnodes.length == 0) ? null : varnodes[0];
+		}
 	}
 
 	/**
 	 * @return last varnode within the ordered list of varnodes
 	 */
 	public Varnode getLastVarnode() {
-		return (varnodes == null || varnodes.length == 0) ? null : varnodes[varnodes.length - 1];
+		Language language = programArch == null ? null : programArch.getLanguage();
+		if (language != null && !language.isBigEndian()) {
+			return (varnodes == null || varnodes.length == 0) ? null : varnodes[0];
+		} else {
+			return (varnodes == null || varnodes.length == 0) ? null : varnodes[varnodes.length - 1];
+		}
 	}
 
 	/**
-- 
2.45.1

