From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: billow <billow.fun@gmail.com>
Date: Sat, 4 May 2024 05:52:33 +0800
Subject: [PATCH] 8780: Fix v850.cspec

---
 Ghidra/Processors/V850/data/languages/V850.cspec | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Ghidra/Processors/V850/data/languages/V850.cspec b/Ghidra/Processors/V850/data/languages/V850.cspec
index 3245283f35..80555113d7 100644
--- a/Ghidra/Processors/V850/data/languages/V850.cspec
+++ b/Ghidra/Processors/V850/data/languages/V850.cspec
@@ -64,7 +64,7 @@
         <register name="sp"/>
         <register name="gp"/>
         <register name="tp"/>
-        <register name="ctbp"/>
+        <register name="CTBP"/>
       </unaffected>
     
     </prototype>
-- 
2.45.1

